In-toto-sign

isachit opened the pull request:
Created the CLI tool to handle the following in context of signing `Link` and `Layout` file:
- add signature, while keeping existing signature(s)
- replace existing signature(s) with new signature
- verify signature(s)

**Issue Referenced** - #97 

Please note the progress:
- [x] Write the script
- [x] Add Exceptions
- [x] Add Documentation
- [x] Write the unit tests and run them
- [x] Fix any other issues, and add other functionalities.

@lukpueh and @SantiagoTorres  can you have a look at it, whenever possible and please let me know if it was what you wanted.

UPDATE 1: in-toto-sign now supports multiple signatures(both multiple signing, and verification of the multiple signatures)
UPDATE 2: users can now specify the destination or target filename of the output of in-toto-sign. In case no destination is specified, the source file name would be taken into consideration by in-toto-sign while dumping.

lukpueh committed:
Adds PKI functions to toto.utils

This commit finally throws away the throw-away-function
'create_and_persist_or_load_key' and replaces it with
'generate_and_write_rsa_keypair' and 'import_rsa_key_from_file'
plus wrappers that additionally prompt for passwords

toto-verify and toto-run are adapted to load keys for signing
and verification using these functions

lukpueh committed:
Split public and private key file contexts

lukpueh committed:
Makes RSA password optional

RSA private keys can now be created and loaded without
password.
Utils figures out whether a password is required when
a key is loaded and only then shows a prompt to the user.
This required some changes to the ssl_crypto submodule
554af0816b4d23eb0e2f8d1bbdabf84372fa3890

lukpueh committed:
Replaces XXX macros with FIXME/TODO

SantiagoTorres committed:
Merge pull request #3 from in-toto/add-docstrings

Adds docstrings everywhere

isachit committed:
Merge Conflict Resolution

isachit committed:
First commit, having code of inspect_sign, Will add comments etc in the coming ones

isachit committed:
Another merge conflict resolution

isachit committed:
Added documentation

isachit committed:
Fixed merge conflicts yet again :'D

isachit committed:
Fixed notations, used argparser's choice method,fixed indentation

isachit committed:
Added link import in verify_sign function

isachit committed:
Removed two unnecessary functions, renamed existing functions, now the helper functions return an object, which is then acted upon as per the user arguments, for example: -i : adds the signature, while dumping infixes the key in the file name; -i -r : adds the signature, replaces all the old signatures, dumps with the above mentioned file name; -r : replaces all the old signatures with the new one, doesn't adds keyid in the file name while dumping, no args - adds the signature to the existing set of signatures, dumps but doesn't adds the keyid while dumping

isachit committed:
Added documentation to the main function

isachit committed:
Added File Header Documentation

isachit committed:
Added action='store_true' in the optional arguments -r (replace all) and -i (infix keyID in the file name)

isachit committed:
Fixed minor documentation issues

isachit committed:
remove in_toto_sign.py from in-toto directory, and moved it to in-toto/in_toto

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+  appends the new signature in the file. Further in each of the two cases,
+  depending on the arguments it then dumps the file either by infixing
+  the keyID i.e. <filename>.<8-initial-characters-from-the-keyID>.link
+  or simpy <filename>.link
```
simply

SantiagoTorres replied:
I think there is also merit in trying to simplify the description too.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+            sys.exit(0)
+
+      except Exception as e:
+        print('The following error occured while signing', e)
```
occurred (also fix below)

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+def add_replace(link, key):
+    """
+    <Purpose>
+      Replaces all the exisiting signature with the new signature,
```
existing

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+        log.logging.getLogger.setLevel(log.logging.INFO)
+
+    if args.operator == 'sign':
+
```
Please remove blank line

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+
+      try:
+        if not args.replaceall:
+
```
Please remove blank line

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+            sys.exit(0)
+
+        else:
+
```
Please remove blank line

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+        sys.exit(2)
+
+    elif args.operator == 'verify':
+
```
Please remove blank line

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+      adding the signature which replaces the old signatures
+    """
+
+    # reading the link file and making an object
```
Please start comments with upper case

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    signable_object = link_import.read_from_file(link)
+
+    # import rsa key from the filepath
+    rsa_key = in_toto.util.import_rsa_key_from_file(key)
```
This won't work if the private key is encrypted. You can use `util.prompt_import_rsa_key_from_file` instead. Also below.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    # check if the key corresponds to the correct format
+    securesystemslib.formats.KEY_SCHEMA.check_match(rsa_key)
+
+    # core working
```
This is not a very useful comment.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    signature = securesystemslib.keys.create_signature(rsa_key, signable_object.payload)
+
+    # append the signature into the file
+    signable_object.signatures.append(signature)
```
Why not use `Signable`'s `sign` method?

isachit replied:
Okay using it.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    # core working
+    signature = securesystemslib.keys.create_signature(rsa_key, signable_object.payload)
+    signable_object.signatures = []
+    signable_object.signatures.append(signature)
```
Why not use `Signable`'s `sign` method (after having removed existing signatures)?

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+      Parsed arguments (args object)
+    """
+    parser = argparse.ArgumentParser(
+        description="in-toto-sign : Signs etc etc")
```
Please make more descriptive.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+        - 'No Signatures Found' - when no signature exists
+
+    <Returns>
+      Boolean True when the verification is success
```
This function does not really have to be a boolean function. Either it raises an exception (something went wrong) or it doesn't (the signature was successfully verified/was correct).

isachit replied:
Alright! Since I am using link.verify_signatures(key_dict), in the function verify_signatures, exceptions are already raised. So what I can do is that, either it returns nothing, the verification was successful, or it catches the exception which occurred while running and prints it . What do you say?

isachit replied:
alright. changed it.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    in_toto_args = parser.add_argument_group("in-toto-sign options")
+
+    in_toto_args.add_argument("-k", "--key", type=str, required=True,
+                              help="Path to private key to sign link metadata (PEM)")
```
Could also be a public key, right?

isachit replied:
Yes. Will change the description.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    rsa_key = in_toto.util.import_rsa_key_from_file(args.key)
+
+    if args.verbose:
+        log.logging.getLogger.setLevel(log.logging.INFO)
```
2 space indentation

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+import securesystemslib.formats
+
+
+def add_sign(link, key):
```
two space indentation (also below)

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    in_toto_args.add_argument("operator", type=str, choices=['sign', 'verify'],
+                              help="sign or verify")
+
+    in_toto_args.add_argument("-r", "--replaceall", action="store_true",
```
Change to `--replace-sig`

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+                              help = "Whether to replace all the old signatures"
+                              "or not")
+
+    in_toto_args.add_argument("-i", "--infixkeyid", action="store_true",
```
Change to `--infix`

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+                              help="sign or verify")
+
+    in_toto_args.add_argument("-r", "--replaceall", action="store_true",
+                              help = "Whether to replace all the old signatures"
```
Please remove whitespaces before and after assignment operator when using keyword arguments consistently throughout the script.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+
+    in_toto_args.add_argument("-i", "--infixkeyid", action="store_true",
+                              help = "whether to infix keyID in the file name"
+                              "or not")
```
Please add a little more detail.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+                    "[--infixkeyid]\n{0}"
+                    "<path/to/signable>"
+                    "[--verbose]\n\n"
+                    .format(lpad))
```
Please review formatting here. Should look more like:

```shell
usage:
in_toto_sign.py (sign | verify) --key <filepath>
                [--replace-sig]
                [--infix]
                [--verbose]
                <path/to/signable>
```

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+                              help="Path to private key to sign link metadata (PEM)")
+
+    in_toto_args.add_argument("operator", type=str, choices=['sign', 'verify'],
+                              help="sign or verify")
```
If you are really motivated and want to learn `argparse` you could use [`subcommands`](https://docs.python.org/2.7/library/argparse.html#sub-commands) for this.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+          else:
+            signable_object = add_replace(args.signablepath, args.key)
+            signable_object.dump(key = rsa_key)
+            sys.exit(0)
```
You could simplify this like so (pseudocode)
```python
try:
  if sign:
    if replace:
       signed_obj = sign_replace(...)

    else:
        signed_obj = sign_add(...)

    if infix:
        signed_obj.dump_with_infix()

    else:
        signed_obj.dump_without_infix()

  else:
     ...
except ExceptionA:
  ...
except ExceptionB:
  ...
...
```


You see what I mean?

isachit replied:
Okay! Yes yes! Changed it!

lukpueh reviewed (changes requested):
Thanks for the PR. Works as expected. Most of my comments are actually style considerations. 

Please also add the `console_script` to `setup.py` and  write unit tests. :)

SantiagoTorres commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+    return signable_object
+
+
+def add_replace(link, key):
```
I'm not sure if this is the best method name. To me it reads as "adding a replace" rather than "replacing any existing signatures"

isachit replied:
replace_sig? What do you say?

SantiagoTorres replied:
sounds good 👍

SantiagoTorres commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+        if verify_sign(args.signablepath, args.key):
+          sys.exit(0)
+        else:
+          sys.exit(1)
```
This will never hit. You either return a `True` or raise an `Exception`, so the condition can never be false.

isachit replied:
changed it.

SantiagoTorres commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,265 @@
+
+      except Exception as e:
+        print('The following error occured while verification', e)
+        sys.exit(3)
```
I"m not sure we are documenting these error codes are we? Is it worth having it something different on each case?

SantiagoTorres reviewed (commented):
Looking good! Thanks for the contrib 👍

isachit committed:
Minor fixes in the documentation, indentation,
Used signable object's sign method for signing
Changed the key import to take into account the case when
the key is encrypted.
Changed the formatting.
Changed some function names.

isachit committed:
Fixed documentation
Remove unnecessary comments

isachit committed:
Added test_in_toto_sign.py .
Edited setup.py
First draft, changes(Additions, deletions etc) to follow

isachit committed:
Missed , in setup.py

isachit committed:
Changed directory of test_in_toto_sign.py to test/test_in_toto_sign.py

isachit committed:
Updated test_in_toto_sign_verify_sign function

isachit committed:
Fixed indentation error, the reason why Travis build was failing

isachit committed:
Changed import names

isachit committed:
Added imports

isachit committed:
Fixed some errors

isachit committed:
Changed key imports

isachit committed:
Added log import in in_toto_sign

isachit committed:
Imported alice's key from file

isachit committed:
Fixed calling of replace-sig, changed function's name to replace_sign

isachit committed:
Same change reflected in the whole file

isachit committed:
Added assertRaises(exceptions.SignatureVerificationError) in the function which verifies the signature
Added missing log in log.logging.getLogger()

isachit committed:
Added self.assertRaises clause in function test_main_required_args

isachit committed:
Added self.assertRaises(SystemExit) in verbose checking

isachit committed:
Fixed documentation issues

isachit committed:
Fixed indentation issues

isachit committed:
Minor fixes in indentation

isachit committed:
Fixed typo in import_rsa_key_from_file

isachit committed:
Removed blank spaces, added comments.
Code was over indented in some places, set it right.

lukpueh commented on test/test_in_toto_sign.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Copy demo files to temp dir
+    for file in os.listdir(demo_files):
+      shutil.copy(os.path.join(demo_files, file), self.test_dir)
```
`file` is reserved in python. Please use a different name.

isachit replied:
I guess file_object should be okay.

lukpueh replied:
IMHO something like "file_path" is better, because `listdir` does not return file objects but paths. ;)

isachit replied:
Okay!

isachit replied:
Done.

lukpueh commented on test/test_in_toto_sign.py:
```diff
@@ -0,0 +1,190 @@
+
+    # Store layout paths to be used in tests
+    self.layout_single_signed_path = "single-signed.layout"
+    self.layout_double_signed_path = "double-signed.layout"
```
You don't need this. It's never used

isachit replied:
Done.

lukpueh commented on test/test_in_toto_sign.py:
```diff
@@ -0,0 +1,190 @@
+    replace_sign(self.layout_single_signed_path, self.alice_path_pvt)
+
+  def test_in_toto_sign_verify_sign(self):
+    """in_toto_sign_verify_sign run through. """
```
This actually does not "run through", right?

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,239 @@
+      verify_sign(args.signablepath, args.key)
+
+  except Exception as e:
+    print('The following error occurred', e)
```
Could you use `log.error` here instead?

isachit replied:
Done.

isachit replied:
Hey @lukpueh , on another note, when asking the user for the target, do he needs to specify a filename? or a path along with the filename to which the file should be dumped?

isachit replied:
For example:

```--destination package_lukas``` 

OR

``` abc/def/lukas/files/important/package_lukas```

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,239 @@
+        sys.exit(0)
+
+    else:
+      verify_sign(args.signablepath, args.key)
```
Could you put this in a separate `try/except` block (not inside this try) and log with `log.fail_verification` on `SignatureVerificationError`, and `log.error` on all other exceptions?

Maybe it would also be nice to log with `log.pass_verification` if the verification passes?

isachit replied:
Done.

isachit replied:
Well, as on here, [common.py](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/common.py) ,It raises SignatureVerificationError on all the cases, no matter if it is due to KeyError, or due to invalid signature or something else.

lukpueh replied:
Yes, `Signable.verify_signatures` raises `SignatureVerificaitonError` whenever something goes wrong.

But your `verify_sign` does not only call `verify_signatures`, it also [loads keys and the file to be verified](https://github.com/isachit/in-toto/blob/c71ccd5152211cb4decab0a25d40371457e977e5/in_toto/in_toto_sign.py#L137-L139). If something goes wrong while doing that I think your tool should `log.error`.

isachit replied:
Ah okay. Cool. :) Added it.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,239 @@
+                            help="Sign or Verify")
+
+  in_toto_args.add_argument("-k", "--key", type=str, required=True,
+                            help="Path to the key ")
```
Do you mind changing this to accept multiple public keys? Currently, `verify` would fail when the file is signed by multiple keys. (Hint: [`nargs='+'`](https://docs.python.org/2.7/library/argparse.html#nargs)) And don't forget to change the usage message. ;)

isachit replied:
Sure. But, while verifying, in case of multiple signatures, the public keys need to be input in the same order as it was signed by the functionaries. That is, for example, the file was first signed by A, then by B (without replacement), then by C (without replacement). When verification takes place, then, pub key of A, followed by B, followed by C, should be specified in the arguments. Right? Or? 

Also there will be multiple cases, depending upon if any one of the functionaries in the supply chain have replaced all the existing signatures or not. What do you say?

isachit replied:
Ah okay. Seems like for verification, the function I have used, takes care of that. It takes input a dictionary of filepaths to various public keys and verifies them. Okay cool.

isachit replied:
![screen shot 2017-06-29 at 6 08 57 pm](https://user-images.githubusercontent.com/10243621/27712674-1abe5540-5cf6-11e7-90f1-56d5a7609408.png)
Any idea why this happens? hm.. Strange..

SantiagoTorres replied:
it's taking the three strings as part of --keys. An easy way out would be to change the parameters to be
```bash
in_toto_sign.py verify <link> --keys key1 [[key2]...] 
```
Or something a litte more git-like
```bash
in_toto_sign verify key1 [[key2]...] -- <path-to-link-file>
```

But I don't know how hard would it be to have this on argparse. Maybe the former is easier.

isachit replied:
Alright! Thank you so much!

lukpueh reviewed (changes requested):
Almost there.

isachit committed:
Added multiple key handling functionality

isachit committed:
Added some code, which was deleted by mistake

isachit committed:
Changed the parameters, edited the tests accordingly

isachit committed:
Missed some code

isachit committed:
Added functions:
1) file_type() : given a file path, it returns 0, if the file is a layout file, returns 1, if it is a link file
2) check_file_type_and_return_object(): given a file path, returns either a link object or a layout object depending upon the file type
3) Added an argument : --d (--destination) which lets the user specify the output file name

isachit committed:
Removed code used for debugging, removed redundant self.assertRaises()

isachit commented:
Why is coverage decreasing. :|

isachit committed:
Removed file_type() function, now checking file type by using signable_object._type

isachit committed:
Removed file_type function, now checking file type by using signable_object._type

isachit committed:
Added functionality which dumps when no key infix is specified

isachit committed:
Removed redundant code, refactored some parts of the code, added a new parameter found in the functon
*function - check_file_type_and_return_object()

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -134,10 +137,28 @@ def verify_sign(link, key_pub):
+          return signable_object
+    if found==-1:
+      raise Exception("Invalid file")
+  file_object.close()
```
We talked about this. Here is what you should do:
Read the file from disk using `open`, try to parse the JSON into a Python dictionary, check if that dict has a key called `_type`, and instantiate the object accordingly using `Layout.read(python_dict)` or `Link.read(python_dict)`.

isachit replied:
Done.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -134,10 +137,28 @@ def verify_sign(link, key_pub):
+          return signable_object
+    if found==-1:
+      raise Exception("Invalid file")
+  file_object.close()
```
Btw. you don't have to close a file explicitly when opened in a `with` block. From the [python docs](https://docs.python.org/2.7/tutorial/inputoutput.html#methods-of-file-objects):

```python
>>> with open('workfile', 'r') as f:
...     read_data = f.read()
>>> f.closed
True
```

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -207,33 +234,87 @@ def main():
+  FILENAME_FORMAT_LAYOUT = "{file_name}.{keyid:.8}.layout"
+  FILENAME_FORMAT_LINK = "{file_name}.{keyid:.8}.link"
+  index=args.signablepath.rfind('/')
+  source_file_name=args.signablepath[index+1:]
```
You could use [`os.path.basename`](https://docs.python.org/2/library/os.path.html#os.path.basename) instead.

lukpueh replied:
Also, try to stick to the codestyle guidelines (whitespace before and after assignment operator).

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -207,33 +234,87 @@ def main():
-  rsa_key = in_toto.util.import_rsa_key_from_file(args.key)
+  length = len(args.keys)
+  rsa_key = in_toto.util.import_rsa_key_from_file(args.keys[length-1])
+  FILENAME_FORMAT_LAYOUT = "{file_name}.{keyid:.8}.layout"
```
I am not sure if we ever need a keyid-infixed layout. But you can leave it.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -207,33 +234,87 @@ def main():
+  length = len(args.keys)
+  rsa_key = in_toto.util.import_rsa_key_from_file(args.keys[length-1])
+  FILENAME_FORMAT_LAYOUT = "{file_name}.{keyid:.8}.layout"
+  FILENAME_FORMAT_LINK = "{file_name}.{keyid:.8}.link"
```
You should import and use the existing  [`FILENAME_FORMAT`](https://github.com/in-toto/in-toto/blob/develop/in_toto/models/link.py#L26) instead.

isachit replied:
Hm. But isn't it just for the link files? I was using it, but then I thought, to make it different for link and layout.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -207,33 +234,87 @@ def main():
   args = parse_args()
-  rsa_key = in_toto.util.import_rsa_key_from_file(args.key)
+  length = len(args.keys)
+  rsa_key = in_toto.util.import_rsa_key_from_file(args.keys[length-1])
```
This comes a little unexpected. I wonder if it wouldn't be better to separate `sign` and `verify` after all, e.g. using a subparser as suggested earlier.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -207,33 +234,87 @@ def main():
+
+          else:
+            signable_object.dump(key=rsa_key)
+            sys.exit(0)
```
Didn't we agree that if no `destination` is specified, you will `dump` overwrite the file? That's not what's happening here.

vladimir-v-diaz replied:
> As a general note, please make sure that unittest coverage does not decrease with your code additions.

Have you tried using coverage's `--fail-under`?  This should make it so that if coverage falls below a certain threshold, a build failure occurs.
```
coverage report -m --fail-under 97
```

I wonder if you can give this option a decimal number...

isachit replied:
@lukpueh Sure. But, what if he specifies, that the keyID must be infixed. True, ```signable_object.dump(key=rsa_key) ``` does not overwrites the file because
1. It takes into account when keyID must be infixed in the file name.
2. Even when it infixes the keyID in the original source file name, it doesn't dump overwrites it since the new file is dumped at a different location than the original source file.

So I guess, the only change that needs to be done, is to dump it at the original location, and overwrite the existing file when `-i` is not specified, and in the other case, remove the existing file, and dump the one with the infixed keyID in it.  Is it what you mean?

isachit replied:
Sure. I will add the tests for those. @vladimir-v-diaz  Thanks!

lukpueh reviewed (changes requested):
The command line arguments here are getting a little confusing. I think we should re-discuss the exact usage of this script. Let me write a follow-up comment that explains what I have in mind before you address any of the comments I wrote in this review.

As a general note, please make sure that unittest coverage does not decrease with your code additions. If you take a look at [`coveralls`' report](https://coveralls.io/builds/12234311/source?filename=in_toto%2Fin_toto_sign.py) you'll see that there are a bunch of lines with red background. These lines are not executed by any unit test, so you would have to add unit tests that trigger that behavior.

lukpueh commented:
Let's take a step back and consider what we actually want here. `in-toto-sign` has two subcommands, `sign` and `verify`. Let's look at each of their options/arguments individually (omitting `-v` because that's straight forward).

**sign**
- `<path/to/input/signable>`
   Path to signable file, must be a link or layout in json format
- `--keys <private key path>[ <private key path>, ...]`
  Create a signature using each key and add each signature to the signable
- `[--replace-sig]` (optional)
  If specified, remove existing signatures before adding new signatures
- `[--destination] <path/to/output/signable>` (optional)
  If specified dump signed link or layout to `<path/to/output/signable>` else dump to `<path/to/input/signable>`
- `[--infix]` (optional)
  Here we have to make some decisions. First, which keyid should we use as infix if the resulting file will have multiple signatures, either because a signature was added to an exiting signature or multiple signatures were added? And second, where should we dump the file to? Let's keep in mind that the keyid infix actually only makes sense for link files signed by one key. I propose the following solutions:
  - Make `--destination` and `--infix` mutually exclusive (fail if both are specified)
  - Ignore `--infix` if the passed signable is a layout (warn user using `log.warn`)
  - Use the key of the last signature that was added as infix (warn user using `log.warn`)
  - Dump the resulting signable to <link_name>.<first 8 digits of keyid>.link in the current working directory (inform user using `log.info`)
  
**verify** (should be straight forward)
- `<path/to/signable>`
- `--keys <public key path>[ <public key path>, ...]`

Let me know would you think about it (especially the `--infix` part)!

isachit committed:
Major changes in code of main function, as suggested by lukas- it now works accordingly and handles all the situations
changed check_file_type_and_return_object() function

isachit commented:
@lukpueh  Please check afb69df77721fdbe0a7e76447010c670bc400673 
I have implemented it according to your last comment and as per our discussion.

isachit committed:
Updated the unit tests, removed blank space in in_toto_sign.py

isachit commented:
@lukpueh  Do you know why the travis CI build isn't happening?

lukpueh commented:
Yes, because there are conflicts in `setup.py`, because we have merged your other PR that has also modified that file.

isachit committed:
Merge branch 'develop' into in-toto-sign

isachit committed:
Typo in test_main_wrong_key_Exits

isachit committed:
Added more tests

isachit committed:
Missed self while defining a path to a link file

isachit committed:
Added more tests

isachit committed:
And more and more tests, edited code of in-toto-sign

isachit committed:
Minor typo

isachit committed:
And more tests

isachit committed:
Edit more tests

isachit commented:
Okay. Well that's nuff. Still -0.2% :| Honestly @coveralls ? Are you kidding me?

isachit commented:
@lukpueh  Sorry to bug you again, but you need to see this.

vladimir-v-diaz commented:
@isachit 
You don't appear to have a test case for line 114 of common.py:
https://coveralls.io/builds/12299985/source?filename=in_toto%2Fmodels%2Fcommon.py

isachit commented:
@vladimir-v-diaz  Thanks maaan! :D Trying to write a test for it.

isachit committed:
Added test for invalid signature verification

isachit committed:
Added different dump files

isachit committed:
Added more tests, when infix and destination both are specified

isachit committed:
Added main's failing verification test

isachit commented:
Yaaye!! :D

isachit commented:
@lukpueh  I guess its done. Just history rewriting remains. Right?

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,326 @@
+  signable_object.verify_signatures(key_pub_dict)
+
+
+def check_file_type_and_return_object(file_path):
```
Can you write a short docstring here?

lukpueh replied:
Can you write a short docstring here?

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,326 @@
+                  "[--verbose]\n{0}"
+                  "[--destination]\n{0}"
+                  "--keys <filepath/filepaths>\n\n"
+                  .format(lpad))
```
Let's just remove the custom usage message (Line 185-195). The default usage message that argparse provides is good enough and more accurate.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,326 @@
+        log.error('Please specify one and only one out of infix and '
+                  'destination!')
+        sys.exit(1)
+      else:
```
You only get to this line if above condition does not match. So you can remove the else here, which makes the code less nested.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,326 @@
+        else:
+          signable_object = add_sign(args.signablepath, args.keys)
+
+        if signable_object._type == 'layout':
```
`os.path.join` only really makes sense if you pass it two arguments, also you should only warn if `--infix` is specified and in general you could make this a little slicker. What do you think about the following?

```python
if args.infix: 
  log.warn('Ignoring option --infix for layouts...')

if args.destination:
  path = args.destination

else:
  path = args.signablepath

log.info("Dumping layout to '{}' ...".format(path))
signable_object.dump(path)
sys.exit(0)
```

isachit replied:
Done.

lukpueh commented on in_toto/in_toto_sign.py:
```diff
@@ -0,0 +1,326 @@
+            log.info('Dumping at ' + str(file_name) + '...')
+            sys.exit(0)
+
+        if signable_object._type == 'Link':
```
Here another suggestion for the link dumping code.

```python
if args.destination:
  path = args.destination

elif args.infix:
  # We will dump the file name to CWD as
  # `<link name>.<keyid of last key passed>.link`
  path = FILENAME_FORMAT_IMPORT.format(
    step_name=signable_object.name, keyid=args.keys[-1]["keyid"])

  if len(args.keys) > 1:
    log.warn("Using last key in the list of passed keys for infix...")

else:
  path = args.signablepath

log.info("Dumping link to '{}'' ...".format(path))
signable_object.dump(path)
sys.exit(0)
```

lukpueh replied:
I guess we could even take it a step further and write the last three lines only once for layouts and links. :)
e.g.:
```python
log.info("Dumping {0} to '{1}'' ...".format(signable_object._type.lower(), path))
signable_object.dump(path)
sys.exit(0)
```

isachit replied:
Ahh okay.

lukpueh reviewed (changes requested):
Awesome coverage! Overall, looking good. Here a couple of final minor suggestions.

isachit committed:
Refactored the code

isachit committed:
Added documentation, removed redundant code

isachit committed:
Edited some tests, minor changed in in_toto_sign.py too

isachit committed:
Fixed an error due to unimported key

isachit committed:
Simplified the code further

isachit commented:
+0.1% Yaaayyee!!
![ ](https://media.giphy.com/media/JltOMwYmi0VrO/giphy.gif)

lukpueh commented:
Thanks @isachit!
